libp2p/module.go

```diff
@@ -0,0 +1,358 @@
+	readStreamTimeoutDuration = time.Second * 10
+	// ErrModule wraps errors which occur within the libp2pModule implementation.
+	ErrModule = typesP2P.NewErrFactory("LibP2P module error")
+)
+
+func Create(bus modules.Bus) (modules.Module, error) {
+	return CreateWithProviders(
+		bus,
+		persistence.NewPersistenceAddrBookProvider(bus),
+		bus.GetConsensusModule(),
+	)
+}
+
+func CreateWithProviders(
+	bus modules.Bus,
+	addrBookProvider addrbook_provider.AddrBookProvider,
+	currentHeightProvider providers.CurrentHeightProvider,
+) (modules.Module, error) {
+	return new(libp2pModule).CreateWithProviders(bus, addrBookProvider, currentHeightProvider)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	// TODO: double check if this should change.
+	return modules.P2PModuleName
+}
+
+func (mod *libp2pModule) Create(bus modules.Bus) (modules.Module, error) {
+	return Create(bus)
+}
+
+func (mod *libp2pModule) CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	*mod = libp2pModule{
+		logger:                new(modules.Logger),
+		addrBookProvider:      addrBookProvider,
+		currentHeightProvider: currentHeightProvider,
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		return nil, ErrModule("unable to register module", err)
+	}
+
+	mod.logger.Print("Creating libp2p-backed network module")
+
+	mod.cfg = bus.GetRuntimeMgr().GetConfig().P2P
+
+	// TODO: support RainTree network
+	if mod.cfg.UseRainTree {
+		return nil, ErrModule("raintree is not yet compatible with libp2p", nil)
+	}
+
+	// TODO(future): investigate any unnecessary
+	// key exposure / duplication in memory
+	secretKey, err := poktCrypto.NewLibP2PPrivateKey(mod.cfg.PrivateKey)
+	if err != nil {
+		// TODO: wrap error
+		return nil, err
+	}
+
+	mod.identity = libp2p.Identity(secretKey)
+
+	switch mod.cfg.ConnectionType {
+	case types.ConnectionType_TCPConnection:
+		addr, err := mod.getMultiaddr()
+		if err != nil {
+			return nil, ErrModule("parsing multiaddr fom config", err)
+		}
+		mod.listenAddrs = libp2p.ListenAddrs(addr)
+	case types.ConnectionType_EmptyConnection:
+		mod.listenAddrs = nil
+	default:
+		return nil, ErrModule("", fmt.Errorf(
+			// DISCUSS: should we refer to this as transport instead?
+			"unsupported connection type: %s", mod.cfg.ConnectionType,
+		))
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		// TODO: wrap error
+		return nil, ErrModule("registering module", err)
+	}
+
+	return mod, nil
+}
+
+func (mod *libp2pModule) Start() error {
+	*mod.logger = logger.Global.CreateLoggerForModule("P2P")
+
+	// IMPROVE: receive context in interface methods?
+	ctx := context.Background()
+
+	// TECHDEBT: metrics integration.
+	var err error
+	opts := []libp2p.Option{
+		mod.identity,
+		mod.listenAddrs,
+		// TECHDEBT / INCOMPLETE: add transport security!
+	}
+
+	if !mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
```

Comment:
```suggestion
        // Disabling some options for the local debug client because ...
	if !mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
```